src/Algebra/Graph/Labelled/AdjacencyMap/Algorithm/Dijkstra.hs

```diff
@@ -0,0 +1,109 @@
+module Algebra.Graph.Labelled.AdjacencyMap.Algorithm.Dijkstra where
+
+import Data.Map.Strict (Map, (!))
+import Algebra.Graph.Label (Semiring(..), (<+>), zero)
+import qualified Data.Map.Strict as Map
+import Algebra.Graph.Labelled.AdjacencyMap.Internal (AdjacencyMap(..))
+
+-- | Psuedocode for Generic-Single-Source-Shortest-Distance (G,s)
+-- for i <- 1 to Number of Vertices:
+--   do d[i] <- r[i] <- 0 
+-- d[s] ← r[s] ← 1 
+-- q ← {s} 
+-- while q != []
+--   do p <- head(q)
+--      Dequeue(q)
+--      r0 <- r[p]
+--      r[p] <- 0
+--      for each e in E[p]
+--        do if d[n[e]] != d[n[e]] <+> (r0 <.> w[e])
+--             then d[n[e]] <- d[n[e]] <+> (r0 <.> w[e])
+--                  r[n[e]] <- r[n[e]] <+> (r0 <.> w[e])
+--                  if n[e] not in q
+--                    then Enqueue(q, n[e])
+-- d[s] ← 1
+--
+-- Convensions used:
+-- if e is a -> b
+-- n[e] = b
+-- p[e] = a
+--
+-- if p is any vertex
+-- E[p] is the edge set of p
+--
+-- if e is any edge
+-- w[e] is the weight of the edge e
+--
+-- Brief:
+-- We use a queue q to maintain the set of vertices whose leaving edges are to be relaxed. q is initialized to [source].
+-- We maintain 2 attributes, namely, `d` and `r`.
+-- d[p] an estimate of the shortest distance from s to p
+-- r[p] is the total weight added to d[q] since the last time p was extracted for q.
+-- We update d and r till the queue is empty.
+
+dijkstra ::
+     (Ord a, Eq b, Semiring b) => a -> AdjacencyMap b a -> Map a b
+dijkstra src adjMap = fst3 fDRQ
+  where
+    iDRQ = initialize src adjMap
+    mDRQ = recursiveUpdateByQ adjMap iDRQ
+    fDRQ = finalize src mDRQ
+    fst3 (x, _, _) = x
+
+initialize ::
+     (Ord a, Semiring b) => a -> AdjacencyMap b a -> (Map a b, Map a b, [a])
+initialize src adjMap = (srcOne, srcOne, [src])
+  where
+    allZero = Map.map (const zero) $ adjacencyMap adjMap
+    srcOne = Map.insert src one allZero
+
+finalize ::
+     (Ord a, Semiring b)
+  => a
+  -> (Map a b, Map a b, [a])
+  -> (Map a b, Map a b, [a])
+finalize s (d, r, queue) = (Map.insert s one d, r, queue)
+
+recursiveUpdateByQ ::
+     (Ord a, Eq b, Semiring b)
+  => AdjacencyMap b a
+  -> (Map a b, Map a b, [a])
+  -> (Map a b, Map a b, [a])
+recursiveUpdateByQ adjMap dRQ@(d, r, queue) =
+  case queue of
+    [] -> dRQ
+    (k:rest) -> recursiveUpdateByQ adjMap $ updateDRQByEdgeMap r' edgeMap cDRQ
+      where r' = r ! k
+            newR = Map.insert k zero r
+            edgeMap = adjacencyMap adjMap ! k
+            cDRQ = (d, newR, rest)
+
+updateDRQByEdgeMap ::
+     (Ord a, Eq b, Semiring b)
+  => b
+  -> Map a b
+  -> (Map a b, Map a b, [a])
+  -> (Map a b, Map a b, [a])
+updateDRQByEdgeMap r' edgeMap pDRQ =
+  foldr (\k dRQ -> updateDRQByEdge r' (k, edgeMap ! k) dRQ) pDRQ $
```

Comment:
Why using `Map.keys` and then doing a `foldr`? Can not this be done by a single traversal using `Map.foldrWithKey`?

Reply:
> Can not this be done by a single traversal using Map.foldrWithKey?

Yes, I'll change that. It is better to use `Map.foldrWithKey`.